### DISCUSSION TITLE
How do I write preload

### DISCUSSION THREAD

**Discussion by tomocrafter** [diesel-rs/diesel#3498] [Q&A] [answered]
There's function `preload` that preloads related models (and function `with` in eloquent).
https://apidock.com/rails/ActiveRecord/QueryMethods/preload

How can I do some thing on diesel?

**Comment by weiznich** [comment, marked as answer]
I do not use the mentioned frameworks myself so I'm not 100% sure how to answer your question. 
You might want to look at [`BelogingToDsl`](https://docs.diesel.rs/2.0.x/diesel/prelude/trait.BelongingToDsl.html#tymethod.belonging_to) and more generally on [`diesel::associations`](https://docs.diesel.rs/2.0.x/diesel/associations/index.html) for similar functionality.

**Reply by tomocrafter** [reply]
Thank you for anwser!

Lets say table schema is:
```Rust
table ! {
  users (id) {
    id -> Uuid,
    detail_id -> Uuid,
  }
}

table ! {
  details (id) {
    id -> Uuid,
    email -> String,
  }
}

table ! {
  posts (id) {
    id -> Uuid,
    user_id -> Uuid,
    body -> String,
  }
}
```

and If using preload in those frameworks, then it fetches data with multiple query like
```
SELECT * FROM users
SELECT * FROM details WHERE id IN (1, 2, 3)
SELECT * FROM posts WHERE id IN (1, 2, 3)
```

then I want to have struct like:
```Rust
struct User {
  id: Uuid,
  detail: Detail,
  posts: Vec<Post>
}

struct Detail {
  id: Uuid,
  email: String
}

struct Post {
  id: Uuid,
  body: String
}
```

**Reply by weiznich** [reply]
That's what the  `diesel::associations` API implements, although some details are different due to rust being a statically typed language. For loading data from your `details` table you want to use an explicit join instead, as that saves another query. I encourage you to checkout the documentation of `diesel::associations` and `diesel::prelude::QueryDsl::inner_join` for details.

**Reply by tomocrafter** [reply]
Thank you so much! I definitely give it try and I'll ask again if we couldn't achive it with that.
